### PR TITLE
[feature] container occupancy detector nodelet

### PR DESCRIPTION
**Pull request by mqcmd196**
The feature for calculate occupancy in containers (bounding boxes).  This PR requires https://github.com/jsk-ros-pkg/jsk_recognition/pull/2695

TODO
- [x] add test
- [x] add doc
~~keep attenion clipper's bounding box labels same~~

### CODE REVIEW

**Comment by mqcmd196** [comment]
@tkmtnt7000 Sorry for being late, I made the release version of `container_occupancy_detector`. If you have time, please test it with `roslaunch jsk_pcl_ros container_occupancy_detector.launch CONTAINER_BOXES_INPUT:=... POINTCLOUD_INPUT:=...` in real environment. `CONTAINER_BOXES_INPUT` is trashibins' boxes topic name, `POINTCLOUD_INPUT:=...` is point cloud topic name

**Comment by tkmtnt7000** [comment]
@mqcmd196 
Thank you! That's a nice feature.
I"ll test this node in my local PC and in Fetch's daily kitchen demo.

File: jsk_pcl_ros/CMakeLists.txt [hunk 1]
```diff
@@ -626,6 +628,7 @@ if (CATKIN_ENABLE_TESTING)
     add_rostest(test/test_voxel_grid_downsample.test)
     add_rostest(test/test_voxel_grid_large_scale.test)
     add_rostest(test/test_organized_statistical_outlier_removal.test)
+    add_rostest(test/test_container_occpancy_detector.test)
```
Comment (tkmtnt7000) [inline jsk_pcl_ros/CMakeLists.txt:631]: Test fails because of file name typo.
`occpancy` ==> `occupancy`

**Review by tkmtnt7000 — CHANGES_REQUESTED** [review]

File: jsk_pcl_ros/CMakeLists.txt (hunk 1 above)
Comment (mqcmd196) [reply jsk_pcl_ros/CMakeLists.txt]: sorry, fixed

**Comment by mqcmd196** [comment]
Sorry for force pushing, I fixed typo in CMakeLists.txt

**Comment by mqcmd196** [comment]
Test failes in
indigo
```cpp
#include <tf2_eigen/tf2_eigen.h>
```

**Comment by tkmtnt7000** [comment]
I confirmed that this feature works well in daily kitchen demo and on my local PC!!

**Comment by mqcmd196** [comment]
@tkmtnt7000 Thank you for your testing!

**Comment by mqcmd196** [comment]
Test passed. Could you review it, maintainers?

File: jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        //     what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
+        // Also see https://github.com/ros/ros_comm/issues/720 .
+        sync_.reset();
+    }
+
+    void ContainerOccupancyDetector::subscribe(){
+        sub_boxes_.subscribe(*pnh_, "container/boxes", 1);
+        sub_points_.subscribe(*pnh_, "container/points", 1);
+        sub_point_indices_.subscribe(*pnh_, "container/point_indices", 1);
+        sync_ = std::make_shared<message_filters::Synchronizer<SyncPolicy> >(100);
```
Comment (iory) [inline jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp:67]: Adding an approximate sync option is a good direction for other use cases.

**Review by iory — CHANGES_REQUESTED** [review]

File: jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp (hunk 2 above)
Comment (mqcmd196) [reply jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp]: Thanks, I added

File: jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    ContainerOccupancyDetector::~ContainerOccupancyDetector(){
+        // message_filters::Synchronizer needs to be called reset
+        // before message_filters::Subscriber is freed.
+        // Calling reset fixes the following error on shutdown of the nodelet:
+        // terminate called after throwing an instance of
+        // 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
+        //     what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
+        // Also see https://github.com/ros/ros_comm/issues/720 .
+        sync_.reset();
```
Comment (iory) [inline jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp:62]:
```suggestion
        sync_.reset();
        ap_sync_.reset();
```

File: doc/jsk_pcl_ros/nodes/container_occupancy_detector.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+Original pointcloud.
+
+* `~container/point_indices` (`jsk_recognition_msgs/ClusterPointIndices`)
+
+The indices of point cloud in each box.
+
+## Publishing Topic
+* `~container/occupancies` (`jsk_recognition_msgs/BoundingBoxArray`)
+
```
Comment (iory) [inline doc/jsk_pcl_ros/nodes/container_occupancy_detector.md:22]: It's good to add `Parameters` section for `queue_size` and `approximate_sync`.
ref: https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/doc/jsk_perception/nodes/apply_mask_image.md#parameters

**Review by iory — CHANGES_REQUESTED** [review]

File: jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                        Eigen::Vector3d rotated_point_ = q_.conjugate() * (eigen_point_ - t_);
+                        float rate_ = (rotated_point_.z() - buttom_.z()) / (top_.z() - buttom_.z()); // remove points under the box
+                        if(rate_ > 0){
+                            sum_occupancy_ += rate_;
+                            n_points_++;
+                        }
+                    }
+                    result_.boxes.at(i_box).value = sum_occupancy_ / float(n_points_);
+                }
+                boxes_occupancy_pub_.publish(result_);
```
Comment (iory) [inline jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp:130]:
```suggestion
                vital_checker_->poke();
                boxes_occupancy_pub_.publish(result_);
```

**Review by iory — CHANGES_REQUESTED** [review]

File: jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp (hunk 3 above)
Comment (mqcmd196) [reply jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp]: Sorry, I forgot ...

File: doc/jsk_pcl_ros/nodes/container_occupancy_detector.md (hunk 4 above)
Comment (mqcmd196) [reply doc/jsk_pcl_ros/nodes/container_occupancy_detector.md:22]: added

File: jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp (hunk 5 above)
Comment (mqcmd196) [reply jsk_pcl_ros/src/container_occupancy_detector_nodelet.cpp:135]: added

**Review by iory — APPROVED** [review]

**Pull request merged**